File: canvas-data/perl/hmac_signature.pl
#!/usr/bin/perl

# This is a code snippet

# Standard modules

use strict;
use warnings;
use diagnostics;
use Carp;

# Specific modules required for signature

use URI;
use HTTP::Date;
use Digest::SHA qw{hmac_sha256_base64};

# Minimal information needed to generate signature

my $timestamp = time2str();
my $url    = 'https://portal.inshosteddata.com/api/account/self/dump?limit=100';
my $api_secret = 'YOUR SUPER SECRET API SECRET GOES HERE';

# Generate signature

my $hmac = hmac_signature( $timestamp, $url, $api_secret );
printf("HMAC signature: %s\n", $hmac);

# Headers to add to request

my $api_key = 'YOUR API KEY GOES HERE';
my $headers = {
  'Authorization' => 'HMACAuth ' . $api_key . ':' . $hmac,
  'Date'          => $timestamp,
};

1;

# Subroutine to generate the proper signature
# This is a whittled down version from what Canvas suggested. 
# Currently, the Content-type and Content-MD5 are not used

sub hmac_signature {
  my $timestamp = shift || return;
  my $url       = shift || return;
  my $secret    = shift || return;
  my $uri       = URI->new($url);
  my $host = defined( $uri->host ) ? $uri->host : 'portal.inshosteddata.com';
  my $query = $uri->query || '';
  if ( $query ne '' ) {
    $query = join( '&', sort( split( '&', $query ) ) );
  }
  my $parts =
    [ 'GET', $host, '', '', $uri->path, $query, $timestamp, $secret, ];
  my $message = join( "\n", @{$parts} );
  my $hmac = hmac_sha256_base64( $message, $secret );

  # The HMAC implementation doesn't enforce the proper length.
  # Pad the end with = signs until the length is a multiple of 4
  while ( length($hmac) % 4 ) {
    $hmac .= '=';
  }
  return $hmac;
}

File: canvas-data/php/hmac_signature.php
<?php
// This is a code snippet

// Minimal information needed to generate signature
$timestamp = gmdate( 'D, d M Y H:i:s T' );
$url = 'https://portal.inshosteddata.com/api/account/self/dump?limit=3';
$api_secret = 'YOUR SUPER SECRET API SECRET GOES HERE';

// Generate signature
$hmac = hmac_signature( $timestamp, $url, $api_secret );
printf( "HMAC signature: %s\n", $hmac );

// Headers to add to request
$api_key = 'YOUR API KEY GOES HERE';
$http_headers = array ( 'Authorization: HMACAuth ' . $api_key . ':' . $hmac, 
    'Date: ' . $timestamp );

// This version is slimmed down to only include what is needed for the current API
function hmac_signature($timestamp = NULL, $url = NULL, $secret = NULL) {
  if (empty( $timestamp ) || empty( $url ) || empty( $secret )) {
    return;
  }
  $u = parse_url( $url );
  if ($u === FALSE) {
    return;
  }
  $host = ! empty( $u['host'] ) ? $u['host'] : 'portal.inshosteddata.com';
  $path = $u['path'];
  $query = '';
  if (! empty( $u['query'] )) {
    $parms = explode( '&', $u['query'] );
    sort( $parms );
    $query = join( '&', $parms );
  }
  $parts = array ( 'GET', $host, '', '', $path, $query, $timestamp, $secret );
  $message = join( "\n", $parts );
  return base64_encode( hash_hmac( 'sha256', $message, $secret, TRUE ) );
}
